common/include/WorkgroupDataAccessors.hlsl

```diff
@@ -0,0 +1,121 @@
+#ifndef _WORKGROUP_DATA_ACCESSORS_HLSL_
+#define _WORKGROUP_DATA_ACCESSORS_HLSL_
+
+namespace nbl
+{
+namespace hlsl
+{
+
+struct ScratchProxy
+{
+    template<typename AccessType, typename IndexType>
+    void get(const uint32_t ix, NBL_REF_ARG(AccessType) value)
+    {
+        value = scratch[ix];
+    }
+    template<typename AccessType, typename IndexType>
+    void set(const uint32_t ix, const AccessType value)
+    {
+        scratch[ix] = value;
+    }
+
+    uint32_t atomicOr(const uint32_t ix, const uint32_t value)
+    {
+        return glsl::atomicOr(scratch[ix],value);
+    }
+
+    void workgroupExecutionAndMemoryBarrier()
+    {
+        glsl::barrier();
+        //glsl::memoryBarrierShared(); implied by the above
+    }
+};
+
+template<class Config, class Binop>
+struct DataProxy
+{
+    using dtype_t = vector<uint32_t, Config::ItemsPerInvocation_0>;
+
+    static DataProxy<Config, Binop> create()
+    {
+        DataProxy<Config, Binop> retval;
+        retval.workgroupOffset = glsl::gl_WorkGroupID().x * Config::VirtualWorkgroupSize;
+        retval.outputBufAddr = sizeof(uint32_t) + pc.pOutputBuf[Binop::BindingIndex];
+        return retval;
+    }
+
+    template<typename AccessType, typename IndexType>
+    void get(const IndexType ix, NBL_REF_ARG(AccessType) value)
+    {
+        value = vk::RawBufferLoad<AccessType>(pc.pInputBuf + (workgroupOffset + ix) * sizeof(AccessType));
+    }
+    template<typename AccessType, typename IndexType>
+    void set(const IndexType ix, const AccessType value)
+    {
+        vk::RawBufferStore<AccessType>(outputBufAddr + sizeof(AccessType) * (workgroupOffset+ix), value, sizeof(uint32_t));
+    }
+
+    void workgroupExecutionAndMemoryBarrier()
+    {
+        glsl::barrier();
+        //glsl::memoryBarrierShared(); implied by the above
+    }
+
+    uint32_t workgroupOffset;
+    uint64_t outputBufAddr;
```

Comment:
this is just convoluted you could just have an input address and output address that has the workgroup offset already added

Reply:
Also did you see https://github.com/Devsh-Graphics-Programming/Nabla/blob/master/include/nbl/builtin/hlsl/bda/legacy_bda_accessor.hlsl#L71-90